sdcm/sct_events/database.py

```diff
@@ -216,3 +210,93 @@ def __init__(self, message: str, severity: Severity = Severity.ERROR):
+    event_class: Type[ContinuousEvent]
+    period_func: Callable
+
+
+class ScyllaDatabaseContinuousEvent(ContinuousEvent, abstract=True):
+    begin_pattern: str = NotImplemented
+    end_pattern: str = NotImplemented
+
+    def __init__(self, node: str, severity=Severity.UNKNOWN, publish_event=False):
+        super().__init__(severity=severity, publish_event=publish_event)
+        self.node = node
+
+
+class ScyllaServerStatusEvent(ScyllaDatabaseContinuousEvent):
+    begin_pattern = r'Starting Scylla Server'
+    end_pattern = r'Stopping Scylla Server'
+
+    def __init__(self, node: str, severity=Severity.NORMAL, publish_event=True):
+        super().__init__(node=node, severity=severity, publish_event=publish_event)
+
+
+class BootstrapEvent(ScyllaDatabaseContinuousEvent):
+    begin_pattern = r'Starting to bootstrap'
+    end_pattern = r'Bootstrap succeeded'
+
+    def __init__(self, node: str, severity=Severity.NORMAL, publish_event=True):
+        super().__init__(node=node, severity=severity, publish_event=publish_event)
+
+
+SCYLLA_DATABASE_CONTINUOUS_EVENTS = [
+    ScyllaServerStatusEvent,
+    BootstrapEvent
+]
+
+
+def get_pattern_to_event_to_func_mapping(event_registry: ContinuousEventsRegistry,
+                                         node: str) \
+        -> List[ScyllaServerEventPatternFuncs]:
+    '''
+    This function maps regex patterns, event classes and begin / end
+    functions into ScyllaServerEventPatternFuncs object. Helper
+    functions are delegated to find the event that should be the
+    target of the start / stop action, or creating a new one.
+    '''
+    mapping = []
+
+    def _add_event(event_type: Type[ContinuousEvent]):
+        new_event = event_type(node=node)
+        new_event.begin_event()
+
+    def _end_event(event_type: Type[ContinuousEvent]):
+        event_filter = event_registry.get_registry_filter()
+        event_filter \
+            .filter_by_node(node=node) \
+            .filter_by_type(event_type=event_type) \
+            .filter_by_period(period_type=EventPeriod.Begin.value)
+
+        begun_events = event_filter.get_filtered()
+
+        if not begun_events:
+            raise ContinuousEventRegistryException("Did not find any events of type {event_type}"
+                                                   "with period type {period_type}."
+                                                   .format(event_type=event_type,
+                                                           period_type=EventPeriod.Begin.value))
+        if len(begun_events) > 1:
+            LOGGER.warning("Found {event_count} events of type {event_type} with period {event_period}. "
+                           "Will apply the function to the oldest event by default."
+                           .format(event_count=len(begun_events),
+                                   event_type=event_type,
+                                   event_period=EventPeriod.Begin.value))
+        event = begun_events[0]
```

Comment:
Let's select the last one: `begun_events[-1]`

Reply:
Fixed (and changed the error message above to reflect this)